<a href="https://colab.research.google.com/github/vrao1/AI-ML/blob/main/TensorFlow/Sales_Forecast_using_TensorFlow.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [23]:
import pandas as pd

Mounting Google Drive to access **datasets** **bold text**

In [24]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [25]:
def loadDataframe():
  train_df = pd.read_csv('/content/drive/MyDrive/DatasetsForGoogleColabWork/TensorFlow/SalesForcast/weekly_sales.csv')
  features_df = pd.read_csv('/content/drive/MyDrive/DatasetsForGoogleColabWork/TensorFlow/SalesForcast/features.csv')
  stores_df = pd.read_csv('/content/drive/MyDrive/DatasetsForGoogleColabWork/TensorFlow/SalesForcast/stores.csv')

  return (train_df, features_df, stores_df)

In [26]:
(t_df, f_df, s_df) = loadDataframe()

In [27]:
print(t_df)

      Store  Dept        Date  Weekly_Sales  IsHoliday
0         1     1  2010-02-05      24924.50      False
1         1     1  2010-02-12      46039.49       True
2         1     1  2010-02-19      41595.55      False
3         1     1  2010-02-26      19403.54      False
4         1     1  2010-03-05      21827.90      False
...     ...   ...         ...           ...        ...
9992      6     8  2010-04-30      47811.70      False
9993      6     8  2010-05-07      46289.02      False
9994      6     8  2010-05-14      46403.86      False
9995      6     8  2010-05-21      46435.02      False
9996      6     8  2010-05-28      47246.38      False

[9997 rows x 5 columns]


In [28]:
print(f_df)

      Store        Date  Temperature  Fuel_Price  MarkDown1  MarkDown2  \
0         1  2010-02-05        42.31       2.572        NaN        NaN   
1         1  2010-02-12        38.51       2.548        NaN        NaN   
2         1  2010-02-19        39.93       2.514        NaN        NaN   
3         1  2010-02-26        46.63       2.561        NaN        NaN   
4         1  2010-03-05        46.50       2.625        NaN        NaN   
...     ...         ...          ...         ...        ...        ...   
8185     45  2013-06-28        76.05       3.639    4842.29     975.03   
8186     45  2013-07-05        77.50       3.614    9090.48    2268.58   
8187     45  2013-07-12        79.37       3.614    3789.94    1827.31   
8188     45  2013-07-19        82.84       3.737    2961.49    1047.07   
8189     45  2013-07-26        76.06       3.804     212.02     851.73   

      MarkDown3  MarkDown4  MarkDown5         CPI  Unemployment  IsHoliday  
0           NaN        NaN        

In [29]:
print(s_df)

    Store Type    Size
0       1    A  151315
1       2    A  202307
2       3    B   37392
3       4    A  205863
4       5    B   34875
5       6    A  202505
6       7    B   70713
7       8    A  155078
8       9    B  125833
9      10    B  126512
10     11    A  207499
11     12    B  112238
12     13    A  219622
13     14    A  200898
14     15    B  123737
15     16    B   57197
16     17    B   93188
17     18    B  120653
18     19    A  203819
19     20    A  203742
20     21    B  140167
21     22    B  119557
22     23    B  114533
23     24    A  203819
24     25    B  128107
25     26    A  152513
26     27    A  204184
27     28    A  206302
28     29    B   93638
29     30    C   42988
30     31    A  203750
31     32    A  203007
32     33    A   39690
33     34    A  158114
34     35    B  103681
35     36    A   39910
36     37    C   39910
37     38    C   39690
38     39    A  184109
39     40    A  155083
40     41    A  196321
41     42    C   39690
42     43  

In [30]:
features_columns = f_df.columns

print("All columns of features dataset = {}".format(features_columns))

store_columns = s_df.columns

print("All columns of store dataset = {}".format(store_columns))

weekly_sales_columns = t_df.columns

print("All columns of weekly sales dataset = {}".format(weekly_sales_columns))

All columns of features dataset = Index(['Store', 'Date', 'Temperature', 'Fuel_Price', 'MarkDown1', 'MarkDown2',
       'MarkDown3', 'MarkDown4', 'MarkDown5', 'CPI', 'Unemployment',
       'IsHoliday'],
      dtype='object')
All columns of store dataset = Index(['Store', 'Type', 'Size'], dtype='object')
All columns of weekly sales dataset = Index(['Store', 'Dept', 'Date', 'Weekly_Sales', 'IsHoliday'], dtype='object')


merging stores and features dataframe on store column

In [31]:
merged_df = f_df.merge(s_df, on='Store')

In [32]:
print(merged_df)

      Store        Date  Temperature  Fuel_Price  MarkDown1  MarkDown2  \
0         1  2010-02-05        42.31       2.572        NaN        NaN   
1         1  2010-02-12        38.51       2.548        NaN        NaN   
2         1  2010-02-19        39.93       2.514        NaN        NaN   
3         1  2010-02-26        46.63       2.561        NaN        NaN   
4         1  2010-03-05        46.50       2.625        NaN        NaN   
...     ...         ...          ...         ...        ...        ...   
8185     45  2013-06-28        76.05       3.639    4842.29     975.03   
8186     45  2013-07-05        77.50       3.614    9090.48    2268.58   
8187     45  2013-07-12        79.37       3.614    3789.94    1827.31   
8188     45  2013-07-19        82.84       3.737    2961.49    1047.07   
8189     45  2013-07-26        76.06       3.804     212.02     851.73   

      MarkDown3  MarkDown4  MarkDown5         CPI  Unemployment  IsHoliday  \
0           NaN        NaN       

Finding columns with NaN values

In [33]:
merged_df_columns = merged_df.columns
merged_df_columns_list = merged_df_columns.tolist()

In [34]:
merged_df_columns_list

['Store',
 'Date',
 'Temperature',
 'Fuel_Price',
 'MarkDown1',
 'MarkDown2',
 'MarkDown3',
 'MarkDown4',
 'MarkDown5',
 'CPI',
 'Unemployment',
 'IsHoliday',
 'Type',
 'Size']

In [35]:
all_na = pd.isna(merged_df)
print(all_na)

      Store   Date  Temperature  Fuel_Price  MarkDown1  MarkDown2  MarkDown3  \
0     False  False        False       False       True       True       True   
1     False  False        False       False       True       True       True   
2     False  False        False       False       True       True       True   
3     False  False        False       False       True       True       True   
4     False  False        False       False       True       True       True   
...     ...    ...          ...         ...        ...        ...        ...   
8185  False  False        False       False      False      False      False   
8186  False  False        False       False      False      False      False   
8187  False  False        False       False      False      False      False   
8188  False  False        False       False      False      False      False   
8189  False  False        False       False      False      False      False   

      MarkDown4  MarkDown5    CPI  Unem

In [38]:
na_features = all_na.any()
print(na_features)

Store           False
Date            False
Temperature     False
Fuel_Price      False
MarkDown1        True
MarkDown2        True
MarkDown3        True
MarkDown4        True
MarkDown5        True
CPI              True
Unemployment     True
IsHoliday       False
Type            False
Size            False
dtype: bool


Count number of null values for each column

In [39]:
print(len(all_na))

8190


In [40]:
print(sum(all_na['MarkDown1']))

4158


In [41]:
print(sum(all_na['MarkDown2']))

5269


In [42]:
print(sum(all_na['MarkDown3']))

4577


In [43]:
print(sum(all_na['MarkDown4']))

4726


In [44]:
print(sum(all_na['MarkDown5']))

4140


In [45]:
print(sum(all_na['CPI']))

585


In [46]:
print(sum(all_na['Unemployment']))

585


All markdown columns values have more or less 50% of entries are missing or null, hence dropping these columns

In [47]:
merged_df.columns

Index(['Store', 'Date', 'Temperature', 'Fuel_Price', 'MarkDown1', 'MarkDown2',
       'MarkDown3', 'MarkDown4', 'MarkDown5', 'CPI', 'Unemployment',
       'IsHoliday', 'Type', 'Size'],
      dtype='object')

In [48]:
all_markdown_columns = ['MarkDown1', 'MarkDown2', 'MarkDown3', 'MarkDown4', 'MarkDown5']
merged_df = merged_df.drop(columns=all_markdown_columns)

merged_df.columns

Index(['Store', 'Date', 'Temperature', 'Fuel_Price', 'CPI', 'Unemployment',
       'IsHoliday', 'Type', 'Size'],
      dtype='object')

CPI and unemployment have 585 << 8190 so we are not dropping